File: src/lib/api/scan.ts
import { invoke } from "@tauri-apps/api/core";
import { toError } from "./errors";
import type { ImportReport, ImportSelection, ScanReport, ScanRequest } from "./types";

// Walks a folder and reports what it found. Registers nothing — the review
// step between this and importScanned is the whole point of the feature.
export async function scanFolder(request: ScanRequest): Promise<ScanReport> {
  try {
    return await invoke<ScanReport>("scan_folder", { request });
  } catch (err) {
    throw toError(err);
  }
}

// Registers the reviewed rows. Best-effort per row: check `failures` on the
// result rather than expecting this to throw for one bad directory.
export async function importScanned(selections: ImportSelection[]): Promise<ImportReport> {
  try {
    return await invoke<ImportReport>("import_scanned", { selections });
  } catch (err) {
    throw toError(err);
  }
}

// The detector kinds this build registers. The scan form's tick-list is built
// from this, never a hardcoded array — a new detector must cost zero frontend
// code.
export async function listDetectorKinds(): Promise<string[]> {
  try {
    return await invoke<string[]>("list_detector_kinds");
  } catch (err) {
    throw toError(err);
  }
}

File: src-tauri/src/commands/scan.rs
use std::sync::Arc;

use tauri::State;

use indexer_core::application::{ImportReport, ImportSelection, ScanService};
use indexer_core::domain::scan::{ScanReport, ScanRequest};
use indexer_core::error::ProjectError;

/// Walks a folder looking for projects the ticked detectors recognize, and
/// returns what it found **without registering anything**. The review step is
/// the point: registering a project is a durable act, and a bulk one should be
/// a deliberate one.
///
/// `async` so the walk runs on Tauri's pool and the window stays responsive.
/// It is blocking from the frontend's point of view — there are no progress
/// events and no cancellation, which is acceptable only because the walk is
/// bounded by `MAX_DIRECTORIES`; `stopped_early` on the report says when that
/// bound was hit.
#[tauri::command]
pub async fn scan_folder(
    scan: State<'_, Arc<ScanService>>,
    request: ScanRequest,
) -> Result<ScanReport, ProjectError> {
    scan.scan(&request)
}

/// Registers the rows the user kept, using the names they left in the review
/// list. Best-effort per row: a directory that fails is reported in
/// `failures` and the rest still land.
#[tauri::command]
pub async fn import_scanned(
    scan: State<'_, Arc<ScanService>>,
    selections: Vec<ImportSelection>,
) -> Result<ImportReport, ProjectError> {
    scan.import(&selections)
}

/// The detector kinds this build registers, so the scan form's tick-list is
/// built from what the binary actually has rather than a hardcoded array —
/// invariant 1, "a new detector is implement + register, zero frontend code".
#[tauri::command]
pub fn list_detector_kinds(scan: State<'_, Arc<ScanService>>) -> Vec<String> {
    scan.detector_kinds()
}
